File: client/src/main/java/se/kth/id2203/kvstore/Console.java
/*
 * The MIT License
 *
 * Copyright 2017 Lars Kroll <lkroll@kth.se>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package se.kth.id2203.kvstore;

import java.io.IOException;
import java.io.PrintWriter;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.apache.log4j.PatternLayout;
import org.apache.log4j.WriterAppender;
import org.jline.reader.EndOfFileException;
import org.jline.reader.LineReader;
import org.jline.reader.LineReaderBuilder;
import org.jline.reader.UserInterruptException;
import org.jline.terminal.Terminal;
import org.jline.terminal.TerminalBuilder;
import util.log4j.ColoredPatternLayout;

/**
 *
 * @author Lars Kroll <lkroll@kth.se>
 */
public class Console implements Runnable {

    private static final String PROMPT = ">";

    private LineReader reader;
    private Terminal terminal;
    private PrintWriter out;
    private final ClientService service;
    private final Map<String, Command> commands = new HashMap<>();
    private final int padTo;

    public Console(ClientService service) {
        this.service = service;
    }

    {
        commands.put("op", new Command() {

            @Override
            public boolean execute(String[] cmdline, ClientService worker) {
                if (cmdline.length == 2) {
                    Future<OpResponse> fr = worker.op(cmdline[1]);
                    out.println("Operation sent! Awaiting response...");
                    try {
                        OpResponse r = fr.get();
                        out.println("Operation complete! Response was: " + r.status);
                        return true;
                    } catch (InterruptedException | ExecutionException ex) {
                        ex.printStackTrace(out);
                        return false;
                    }

                } else {
                    return false;
                }
            }

            @Override
            public String usage() {
                return "op <key>";
            }

            @Override
            public String help() {
                return "Just a test operation...replace with proper put get";
            }
        });
        commands.put("help", new Command() {

            @Override
            public boolean execute(String[] cmdline, ClientService worker) {
                //out.println("Interface currently connected to " + bootstrapAddr + ":" + bootstrapPort + ".\n\n");
                out.println("Available commands: \n\n");
                Set<Command> comSet = new HashSet<>(commands.values());
                StringBuilder sb = new StringBuilder();
                for (Command c : comSet) {
                    sb.append(c.usage());
                    for (int i = c.usage().length(); i < padTo; i++) {
                        sb.append(' ');
                    }
                    sb.append(c.help());
                    sb.append('\n');
                }
                out.println(sb.toString());
                return true;
            }

            @Override
            public String usage() {
                return "help";
            }

            @Override
            public String help() {
                return "shows this help";
            }
        });
        Command exitcom = new Command() {

            @Override
            public boolean execute(String[] cmdline, ClientService worker) {
                out.println("Exiting...");
                System.exit(0);
                return true; // clearly an unreachable statement, mr java compiler -.-
            }

            @Override
            public String usage() {
                return "exit|quit";
            }

            @Override
            public String help() {
                return "closes the shell";
            }
        };
        commands.put("exit", exitcom);
        commands.put("quit", exitcom);
        int longestCom = 0;
        Set<Command> comSet = new HashSet<Command>(commands.values());
        for (Command c : comSet) {
            int useLength = c.usage().length();
            if (useLength > longestCom) {
                longestCom = useLength;
            }
        }
        padTo = longestCom + 4;
    }

    @Override
    public void run() {
        try {
            terminal = TerminalBuilder.terminal();
            reader = LineReaderBuilder.builder()
                    .terminal(terminal)
                    .build();
            out = terminal.writer();
            PatternLayout layout = new ColoredPatternLayout("%d{[HH:mm:ss,SSS]} %-5p {%c{1}} %m%n");
            Logger rootL = LogManager.getRootLogger();
            rootL.removeAllAppenders();
            rootL.addAppender(new WriterAppender(layout, out));
            while (true) {
                String line = null;
                try {
                    line = reader.readLine(PROMPT, null, null, null);
                } catch (UserInterruptException e) {
                    // Ignore
                } catch (EndOfFileException e) {
                    return;
                }
                if (line == null) {
                    continue;
                }
                line = line.trim();
                if (line.isEmpty()) {
                    continue;
                }
                String[] cmdline = line.split(" ", 2);
                String cmd = cmdline[0];
                Command c = commands.get(cmd);
                if (c == null) {
                    cmd = cmd.toLowerCase();
                    c = commands.get(cmd);
                }
                if (c != null) {
                    if (!c.execute(cmdline, service)) {
                        out.print("Usage: ");
                        out.println(c.usage());
                    }
                } else {
                    out.println("Unkown command: " + cmd + " (use 'help' to see available commands)");
                }
            }
        } catch (IOException ex) {
            throw new RuntimeException(ex);
        }
    }

    public static abstract class Command {

        public abstract boolean execute(String[] cmdline, ClientService worker);

        public abstract String usage();

        public abstract String help();
    }
}

File: client/src/main/scala-2.11/kvstore/Console.scala
package kvstore

import java.io.PrintWriter

import com.google.common.util.concurrent.SettableFuture
import org.apache.log4j.{LogManager, Logger, PatternLayout, WriterAppender}
import org.jline.reader.{LineReader, LineReaderBuilder}
import org.jline.terminal.{Terminal, TerminalBuilder}
import stormy.kv.OperationResponse

import scala.collection.mutable
import scala.concurrent.TimeoutException
import scala.concurrent.duration.TimeUnit

// TODO Clean this messy thing up!

object Console {
    private final val PROMPT: String = "> "
    private final val OP_TIMEOUT: Long = 3000
    private final val OP_TIMEOUT_UNIT: TimeUnit = java.util.concurrent.TimeUnit.MILLISECONDS
}

class Console(service: ClientService) extends Runnable {
    import kvstore.Console._

    val exitCommand: Command = new Command() {
        override def execute(cmdline: Array[String], worker: ClientService): Boolean = {
            out.get.println("Exiting...")
            System.exit(0)
            true
        }

        override def usage: String = "exit"

        override def help: String = "Closes the shell"
    }
    private val commands = new mutable.HashMap[String, Command]
    var longestCom: Int = 0

    commands.put("status", new Command() {
        override def execute(cmdline: Array[String], worker: ClientService): Boolean = {
            if (cmdline.length == 2) {
                val fr = worker.op(cmdline)
                handleResponse(fr)
            } else {
                false
            }
        }

        override def usage: String = "status <key>"

        override def help: String = "Requests the partition table status and key location information."
    })
    commands.put("get", new Command() {
        override def execute(cmdline: Array[String], worker: ClientService): Boolean = {
            if (cmdline.length == 2) {
                val fr = worker.op(cmdline)
                handleResponse(fr)
            } else {
                false
            }
        }

        override def usage: String = "get <key>"

        override def help: String = "Attempts to recover the value held at <key>"
    })
    commands.put("put", new Command() {
        override def execute(cmdline: Array[String], worker: ClientService): Boolean = {
            if (cmdline.length == 3) {
                val fr = worker.op(cmdline)
                handleResponse(fr)
            } else {
                false
            }
        }

        override def usage: String = "put <key> <value>"

        override def help: String = "Stores <value> at <key>"
    })
    commands.put("cas", new Command() {
        override def execute(cmdline: Array[String], worker: ClientService): Boolean = {
            if (cmdline.length == 4) {
                val fr = worker.op(cmdline)
                handleResponse(fr)
            } else {
                false
            }
        }

        override def usage: String = "cas <key> <refValue> <newValue>"

        override def help: String = "Performs a compare-and-swap for <key>. If the value at <key> equals <refValue>, <newValue> is stored."
    })
    commands.put("help", new Command() {
        override def execute(cmdline: Array[String], worker: ClientService): Boolean = {
            out.get.println("Available commands: \n\n")
            val comSet= commands.values
            val sb = new StringBuilder()
            for (c <- comSet) {
                val usage = c.usage
                sb.append(usage)
                for (i <- usage.length until padTo) {
                    sb.append(" ")
                }
                sb.append(c.help).append("\n")
            }
            out.get.println(sb.toString())
            true
        }

        override def usage: String = "help"

        override def help: String = "Shows this help"
    })


    var padTo = 0

    commands.put("exit", exitCommand)

    val commandSet: Set[Command] = commands.values.toSet


    private var out: Option[PrintWriter] = None
    private var terminal: Terminal = _
    private var reader: LineReader = _
    for (command <- commandSet) {
        val useLength: Int = command.usage.length
        if (useLength > longestCom) {
            longestCom = useLength
        }
        padTo = longestCom + 4
    }

    override def run(): Unit = {
        terminal = TerminalBuilder.terminal
        reader = LineReaderBuilder.builder()
            .terminal(terminal)
            .build()
        out = Some(terminal.writer())
        val layout: PatternLayout = new PatternLayout()
        // TODO Make coloredPatternLayout
        val rootLogger: Logger = LogManager.getRootLogger
        rootLogger.removeAllAppenders()
        rootLogger.addAppender(new WriterAppender(layout, out.get))
        while (true) {
            var line: Option[String] = None
            try {
                line = Some(reader.readLine(Console.PROMPT, null, null, null))
            } catch {
                case _: Throwable => // Ignore
            }
            if (line.isDefined) {
                line = Some(line.get.trim)
                if (!line.get.isEmpty) {
                    val cmdline = line.get.split(" ")
                    var cmd = cmdline(0)
                    var c = commands.get(cmd)
                    if (c == null) {
                        cmd = cmd.toLowerCase
                        c = commands.get(cmd)
                    }
                    if (c.isDefined ) {
                        if (!c.get.execute(cmdline, service)) {
                            out.get.print("Usage: ")
                            out.get.println(c.get.usage)
                        }
                    }
                    else out.get.println("Unknown command: " + cmd + " (use 'help' to see available commands)")
                }
            }
        }
    }

    private def handleResponse(fr: SettableFuture[OperationResponse]): Boolean = {
        try {
            val resp: OperationResponse = fr.get(OP_TIMEOUT, OP_TIMEOUT_UNIT)
            out.get.println(s"[${resp.status}] ${resp.content.getOrElse("")}")
        } catch {
            case _: TimeoutException =>
                out.get.println(s"[Timeout] Request timed out. System may be reconfiguring.")
            case other: Exception =>
                out.get.println(s"[Warn] Request resulted in exception: ${other.getMessage}")
        }
        true
    }

    abstract class Command {
        def execute(cmdline: Array[String], worker: ClientService): Boolean

        def usage: String

        def help: String
    }

}
